rclcpp_lifecycle/src/transition.cpp

```diff
@@ -36,54 +50,109 @@ Transition::Transition(uint8_t id, const std::string & label)
+  }
   start_state->id = start.id();
-  start_state->label = start.label().c_str();
+  start_state->label =
+    rcutils_strndup(start.label().c_str(), start.label().size(), allocator_);
 
-  auto goal_state = new rcl_lifecycle_state_t;
+  auto goal_state = reinterpret_cast<rcl_lifecycle_state_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_state_t), allocator_.state));
+  if (!goal_state) {
+    throw std::runtime_error("failed to allocate memory for rcl_lifecycle_state_t");
+  }
   goal_state->id = goal.id();
-  goal_state->label = start.label().c_str();
+  goal_state->label =
+    rcutils_strndup(goal.label().c_str(), goal.label().size(), allocator_);
 
   transition_handle->start = start_state;
   transition_handle->goal = goal_state;
   transition_handle_ = transition_handle;
 }
 
-Transition::Transition(const rcl_lifecycle_transition_t * rcl_lifecycle_transition_handle)
-: owns_rcl_transition_handle_(false)
+Transition::Transition(
+  const rcl_lifecycle_transition_t * rcl_lifecycle_transition_handle,
+  rcutils_allocator_t allocator)
+: allocator_(allocator),
+  owns_rcl_transition_handle_(false),
+  transition_handle_(nullptr)
 {
-  transition_handle_ = rcl_lifecycle_transition_handle;
+  transition_handle_ = const_cast<rcl_lifecycle_transition_t *>(rcl_lifecycle_transition_handle);
 }
 
 Transition::~Transition()
 {
-  if (owns_rcl_transition_handle_) {
-    if (transition_handle_->start) {
-      delete transition_handle_->start;
+  if (!transition_handle_) {
+    owns_rcl_transition_handle_ = false;
+    return;
+  }
+
+  // can't free anything which is not owned
+  if (!owns_rcl_transition_handle_) {
+    transition_handle_ = nullptr;
+    return;
+  }
+
+  if (transition_handle_->start) {
+    if (transition_handle_->start->label) {
+      allocator_.deallocate(
+        const_cast<char *>(transition_handle_->start->label), allocator_.state);
+      transition_handle_->start->label = nullptr;
     }
-    if (transition_handle_->goal) {
-      delete transition_handle_->goal;
+    allocator_.deallocate(transition_handle_->start, allocator_.state);
+    transition_handle_->start = nullptr;
+  }
+  if (transition_handle_->goal) {
+    if (transition_handle_->goal->label) {
+      allocator_.deallocate(
+        const_cast<char *>(transition_handle_->goal->label), allocator_.state);
+      transition_handle_->goal->label = nullptr;
     }
-    delete transition_handle_;
+    allocator_.deallocate(transition_handle_->goal, allocator_.state);
+    transition_handle_->goal = nullptr;
+  }
+  if (transition_handle_->label) {
+    allocator_.deallocate(
+      const_cast<char *>(transition_handle_->label), allocator_.state);
+    transition_handle_->label = nullptr;
   }
+  allocator_.deallocate(transition_handle_, allocator_.state);
+  transition_handle_ = nullptr;
 }
 
 uint8_t
 Transition::id() const
 {
+  if (!transition_handle_) {
+    throw std::runtime_error("internal transition_handle is null");
+  }
```

Comment:
If the constructor protects the class from passing `null` explicitly this can never happen.